### DISCUSSION TITLE
authorization and cross-broker bridge

### DISCUSSION THREAD

**Discussion by dberardo-com** [emqx/emqx#12731] [Q&A]
## per-listener-based authorization

i am struggling to understand EMQX documentation when it comes to listeners and authorization.

in particular i would like to disable all listeners but 2: "tcp" and "ws".

the tcp listener should use TLS and the ws should not

tcp listener should use certificate based authentication and authorize using acl_file_1

ws should use password based authentication and authorized using acl_file_2.

anonymous clients are not allowed and should be disconnected when they attempt connection
tcp clients providing a certificate should not be required to provide a username as well, as this is to be extracted from the certificate CN field automatically

## cross-broker bridge

i need to create a bridge from a mosquitto broker to emqx. is it possible to achieve this from the "mosquitto-side", i.e. without any special emqx configuration ?

---
# UPDATE

## Conclusions

1. per-listener authorization -> not yet possible
2. per-listener authentication  --> possible. disabling auth on a listener makes it possible to impersonate just any user
3. declarative definition of usernames/password --> coming soon
4. certificate-based username authentication possible by using different zones (needed when using multiple listeners)

## To Test

- [ x ] test point 4 - https://gist.github.com/zmstone/1c85eeaae20f7f5745ee92e288921b0e
- [ x] test enable_auth = false on ws listener
- [ x] test admin ACL

**Comment by zmstone** [comment]
Hi @dberardo-com 

"tcp listener with TLS" is called `ssl` listener in EMQX configuration terms.
To avoid confusion, I will refer to it as `ssl` listener below.

## Per-listener authentication
There is per listener authentication configuration.
you can find the doc here: https://docs.emqx.com/en/enterprise/v5.4.1/hocon/#V-listeners
Setting `enable_auth` to `false` should turn off authentication for the clients connecting to this listener.

## Per-listener authorization (ACL)

There is however NOT a per-listener authorization configuration.
As a workaround, you may use clientid or username patterns to identify clients connecting to different listeners,
and apply a `re` rule to match clientid or username.

## Quick deny anonymous clients
> anonymous clients are not allowed and should be disconnected when they attempt connection

For this, you should set `enable_auth` to `quick_deny_anonymous` for ws listener.

## Use TLS certificate CommonName as username when there are multiple listeners
> tcp clients providing a certificate should not be required to provide a username

This is simpler when you have only SSL listeners, since you have a WS listener too,
you cannot use the global `mqtt` config (which is implicitly bound to the `default` zone).

You'll need to configure another `zone`, and  bind the SSL listener to the zone,
this is currently not undocumented, but I have an example here:
https://gist.github.com/zmstone/1c85eeaae20f7f5745ee92e288921b0e

**Reply by dberardo-com** [reply]
thanks for  the quick reply.

> There is however NOT a per-listener authorization configuration. 

i kind of figured out that myself too from the doc, but i wanted to go sure.

> you cannot use the global mqtt config (which is implicitly bound to the default zone).

i will try this out tomorrow and letting you know

> As a workaround, you may use clientid or username patterns to identify clients connecting to different listeners,
and apply a re rule to match clientid or username.

i see. do you have a link to any doc related to this point? also: it is possible to use an ACL of this kind, i have got some errors with this config (basically authenticated users :can only interact with their topics, except admin that has globla access) 

```
{allow, {user, "admin"}, all, ["$SYS/#", "#"]}.
{allow, all, all, ["${username}/#"]}.
```

and finally: i am working in k8s, could be possible to add password files as secrets ? basically like mosquitto passwd_file. so the question is: can username/pwd credentials be defined in a declarative waz or must it be done via the CLI in an imperative way ?


> cross-broker bridge

seems like no extra configuration is necessary to interact with a regular mosquitto "downstream" bridge. will confirm that tomorrow

**Comment by zmstone** [comment]
ACL rules are described here:
https://docs.emqx.com/en/enterprise/latest/access-control/authz/file.html
Example: {clientid, {re, "^dash"}} : clients with client ID matching the regular expression "^dash"

> i have got some errors with this config

I cannot spot an error from your rules. Please share which errors you found, so we can help to identify the issue.

> i am working in k8s, could be possible to add password files as secrets ? 

There is currently no support to bootstrap MQTT client username password file into the database, this is planed feature though.
To import users, you can call this admin API:
https://docs.emqx.com/en/enterprise/v5.5/admin/api-docs.html#tag/Authentication/paths/~1authentication~1%7Bid%7D~1import_users/post

> cross-broker bridge

If you want to configure your bridge as EMQX to act as a client and mosquitto as the server, it's a good idea to check the "bridge mode" box.
If you configure mosquitto as the client and EMQX as the server, then there is nothing to configure for EMQX.

**Reply by dberardo-com** [reply]
> There is currently no support to bootstrap MQTT client username password file into the database

mm i see ... so if i am willing to run EMQX as a stateless service, the only way i currently have to handle users is either via an external privder (e.g. LDAP) or via certificates, right ?

> I cannot spot an error from your rules

will give it a try again soon

> Setting enable_auth to false should turn off authentication for the clients connecting to this listener

now, given that i want my ACL above to work for admin users, could i disable auth on one "private listener" so that any client can simply impersonate an admin user by providing just any username/password combination? this could be a temporary solution while waiting for the declarative password file feature to be released, this way one can setup all needed ACLs for all users and impersonating them where needed.

**Reply by dberardo-com** [reply]
ok i have managed to make all tests above and everything seems to work now (under the limitation that authorization per-listener is not available yet and neither do password files).

i have noticed 1 undesired behaviors that make it hard for mosquitto brokers to open their bridge with the EMQX broker:

if one uses the mosquitto bridge notation: 
topic # out 1 
topic # in 0
then EMQX denies access completely to the bridge, which according to the ACL is only authorized to "username/#". is it possible to change this behavior of EMQX so as to make it compatible to mosquitto bridges ?

**Reply by zmstone** [reply]
>  i am willing to run EMQX as a stateless service, the only way i currently have to handle users is either via an external privder (e.g. LDAP) or via certificates, right ?

Correct. You can still call the management API to import after EMQX is boot.

Turning authentication off for a listener would require other means to make it secure
for example:
- firewall rules to allow only trusted source net
- mTLS authentication

**Reply by dberardo-com** [reply]
> Correct. You can still call the management API to import after EMQX is boot.

sure, but since emqx is running in a ephemeral container, i would need to perform the import in an automated way every time it goes back up ... let me try to have a look at the EMQX kubernetes operator, perhaps it already has something like this

btw: do you know if it is possible for EMQX to get the ProxyProtocol working on a TLS passthrough connection via a reverse proxy ? according to this it should be possible: https://www.emqx.io/docs/en/latest/access-control/authz/authz.html#:~:text=EMQX%20supports%20Proxy%20Protocol%2C%20that,only%20applicable%20to%20TLS%20connections. 
, but when i set the proxy protocol configuration v1 or v2 on the reverse proxy i get this error on the client side:

OpenSSL Error[0]: error:0A0003F2:SSL routines::sslv3 alert unexpected message

**Reply by zmstone** [reply]
did you configure `proxy_protocol = true` for the listener?
https://www.emqx.io/docs/en/v5.5.1/hocon/#V-listeners-S-listeners-tcp-S-mqtt_tcp_listener-proxy_protocol

**Reply by dberardo-com** [reply]
i see ... now the error has changed, the client gets and "unauthorized" while the broker outputs this error. perhaps the problem is that i am using Traefik as a reverse proxy and that is not supported yet ?

> errorContext: connection_shutdown, reason: {invalid_proxy_info,<<22,3,1,1,60,1,0,1,56,3,3,36,5,34,84,54,164,218,106,223,115,123,229,91,10>>}, offender: [{pid,<0.3452.0>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => quick_deny_anonymous,limiter => #{client => #{messages => #{burst => 0,divisible => true,failure_strategy => force,initial => 0,low_watermark => 0,max_retry_time => 3600000,rate => 100.0}}},listener => {ssl,certhash},zone => certhash}]}}

**Reply by zmstone** [reply]
EMQX can only work with PPV2 protocol

**Reply by dberardo-com** [reply]
mm i have tried with both versions, do you mean proxyProtocol v2 ? https://doc.traefik.io/traefik/routing/services/#proxy-protocol

**Reply by zmstone** [reply]
yes, proxy protocol v2.

**Reply by dberardo-com** [reply]
exactly, and traefik does implement it, nevertheless i keep getting the error above with both v1 and v2. please notice that i am terminating TLS on the EMQX broker and the proxy is doing tls passthrough besides PPV2. could this be the issue?

**Reply by id** [reply]
@dberardo-com could you share your full Traefik configuration?

**Reply by dberardo-com** [reply]
it is literally as simple as this (using an IngressRouteTCP object):

![image](https://github.com/emqx/emqx/assets/65530457/c9559d24-9468-40eb-8263-471b54e3734b)

**Reply by id** [reply]
Looks legit. Could you do tcpdump on emqx node to trace incoming packets from traefik?

**Reply by dberardo-com** [reply]
i am going to do this soon. will keep you posted.

in the meantime i have some questions on the Limiter functionality. should i post it on a separate thread or use this one ?

**Comment by zmstone** [comment]
> then EMQX denies access completely to the bridge

maybe you  need to `allow` mosquitto client to subscribe `#` topic like this:
`{allow, {uername, "mosquitto-bridge-client"}, subscribe, [{eq, "#"}]}.`

**Reply by dberardo-com** [reply]
let me investigate this. here i am talking about "autodiscovery of topics". but let me make some tests first and will get back to you with the results.
